### ISSUE TITLE
Multiple components of the same type

### ISSUE THREAD

**Issue by simonvpe** [SuperV1234/ecst#26]
I am trying to figure out how one would implement multiple instances of a component. Is having a component like `struct MyComponent { std::vector<MyComponentImpl> impl; }` the suggested way to do this or do you have any other suggestion?

**Comment by OvermindDL1** [comment]
The definition of a component storage is that an entity belongs to it, not belongs to it multiple times.  I'd have the internal data store be an array or list or so depending on what is needed, or change the problem to make that not necessary (I hate hate hate allocating dynamic memory in components).  But yes, your vector bit is 'fine' but not efficient.  Better question is, what is `MyComponentImpl` and what precisely are you wanting to store?

**Comment by simonvpe** [comment]
@OvermindDL1 To describe my particular problem further; I am making a simulation environment for Magic The Gathering (the trading card game). I am modeling cards as entities with components Toughness, Power, Loyalty, ...etc. Cards also have zero or more Abilities which can be activated. When activated they spawn an entity with an Effect component, the EffectSystem then proceeds to call a function pointer on all the active Effects each process loop.

**Comment by OvermindDL1** [comment]
So it is the Abilities then?  For those I'd probably make an Ability an entity that is parented to a card, just duplicate a base Ability blueprint for each one you attach.

**Comment by simonvpe** [comment]
@OvermindDL1 Yes, the Abilitiea. What do you mean by parented to? Parent as in keeping a reference to an entity? In that case, is it safe to keep Entity references between update loop steps?

**Comment by OvermindDL1** [comment]
> Parent as in keeping a reference to an entity? In that case, is it safe to keep Entity references between update loop steps?

I'd have a component on the 'instanced as an entity' ability entity that holds a handle to a card entity.  I'm unsure about this ECS library (I have my own) but a handle is a 32 or 64 bit in my system, where some count of lower bits is the entity index and the rest of the higher bits is the generation id to invalidate it, so it is safe in my system at least, I'd imagine it would be similar here considering this is the handle definition here:  https://github.com/SuperV1234/ecst/blob/master/include/ecst/context/entity/handle/handle.hpp#L39-L40

**Comment by vittorioromeo** [comment]
@simonvpe 

>  is it safe to keep Entity references between update loop steps?

If you are sure that the entity wasn't destroyed during the previous update step, you can refer to it through an `entity_id` safely.

If you are unsure about the state of the entity, you should use an [`entity::handle`](https://github.com/SuperV1234/ecst/blob/d54d7cec72f0020f04fa738bd8088c317c31a33b/include/ecst/context/entity/handle/handle.hpp), which can be queried before access with `context::alive(const handle&)`.
